src/zenml/model/model_version.py

```diff
@@ -0,0 +1,521 @@
+)
+from uuid import UUID
+
+from pydantic import BaseModel, PrivateAttr, root_validator
+
+from zenml.enums import ModelStages
+from zenml.exceptions import EntityExistsError
+from zenml.logger import get_logger
+
+if TYPE_CHECKING:
+    from zenml.models import (
+        ArtifactResponseModel,
+        ModelResponseModel,
+        ModelVersionResponseModel,
+        PipelineRunResponseModel,
+    )
+
+logger = get_logger(__name__)
+
+
+class ModelVersion(BaseModel):
+    """ModelVersion class to pass into pipeline or step to set it into a model context.
+
+    name: The name of the model.
+    license: The license under which the model is created.
+    description: The description of the model.
+    audience: The target audience of the model.
+    use_cases: The use cases of the model.
+    limitations: The known limitations of the model.
+    trade_offs: The tradeoffs of the model.
+    ethics: The ethical implications of the model.
+    tags: Tags associated with the model.
+    version: The model version name, number or stage is optional and points model context
+        to a specific version/stage. If skipped new model version will be created.
+    save_models_to_registry: Whether to save all ModelArtifacts to Model Registry,
+        if available in active stack.
+    """
+
+    name: str
+    license: Optional[str]
+    description: Optional[str]
+    audience: Optional[str]
+    use_cases: Optional[str]
+    limitations: Optional[str]
+    trade_offs: Optional[str]
+    ethics: Optional[str]
+    tags: Optional[List[str]]
+    version: Optional[Union[ModelStages, int, str]]
+    save_models_to_registry: bool = True
+
+    suppress_class_validation_warnings: bool = False
+    was_created_in_this_run: bool = False
+
+    _model_id: UUID = PrivateAttr(None)
+    _id: UUID = PrivateAttr(None)
+    _number: int = PrivateAttr(None)
+
+    #########################
+    #    Public methods     #
+    #########################
+    @property
+    def all_model_versions(self) -> List["ModelVersion"]:
+        """Get all model versions for current model from the Model Control Plane.
+
+        Returns:
+            A list of ModelVersion objects registered within current model.
+        """
+        return self._get_or_create_model().versions
+
+    @property
+    def id(self) -> UUID:
+        """Get version id from  the Model Control Plane.
+
+        Returns:
+            ID of the model version or None, if model version
+                doesn't exist and can only be read given current
+                config (you used stage name or number as
+                a version name).
+        """
+        if self._id is None:
+            try:
+                self._get_or_create_model_version()
+            except RuntimeError:
+                logger.info(
+                    f"Model version `{self.version}` doesn't exist "
+                    "and cannot be fetched from the Model Control Plane."
+                )
+        return self._id
```

Comment:
How about you introduce a property `self.response` with a corresponding attribute `_response` that directly holds the `VersionResponseModel` linked to this class. Then you could have all the fetching logic in one place, you wouldn't need `_id`, `_model_id`, `_number`, and you could simply do return `self.response.id` here ( and similarly `self.response.stage` / `self.response.number` / ... in other methods below)

Reply:
My view is pretty much the same as above: the idea of this PR is to simplify the experience and avoid forcing users to learn about our response models. Model Version should clearly have an `id`, `stage`, and `number`. The user shouldn't worry about how I got them. Storing the whole `_response` instead and fetching from it might be the case, but does it improve the code much? For instance `id` and `number` are immutable entities, while `stage` should be fetched each time it is requested.

Reply:
I agree, it's not user-facing, but I think it would be cleaner since it would reduce code duplication.

Essentially this is the same design pattern that @bcdurak is facing with his model metadata in #1971: You have class A that wraps class B and you want to expose all attributes B.x directly on A as A.x. @bcdurak solved it right now by holding A.b internally and defining A.x as properties that simply return A.b.x, which would also be my suggestion here. However, as I commented on #1971, I think that is not an ideal solution either and I would love if someone came up with an approach that required even less code duplication. 

Anyway, you don't necessarily need to do that as part of this PR, just wanted to expose you to this problem as well, since maybe you'll get an idea how to design this more cleanly (which could perhaps allow us to get rid of several thousand lines of useless code across the ZenML code base as a whole)